README.md
# Measured Elasticsearch Reporter

Reports [measured][] metrics to [elasticsearch][] in the same format as the
[Java elasticsearch metric reporter][java].

## Install

    npm install measured elasticsearch measured-elasticsearch

## Usage

```js
var elasticsearch = require('elasticsearch');
var client        = new elasticsearch.Client({ host : 'localhost:9200' });

var reporter   = require('measured-elasticsearch').forClient(client);
var measured   = require('measured');
var collection = measured.createCollection();

reporter.addCollection(collection);
reporter.start(60, measured.units.SECONDS);
```

## API

- `forClient(client[, config])` creates a new reporter for the given client and
  optional config. Available configs are:
    - `index` the index to use. Defaults to `metrics`.
    - `indexDateFormat` the date format to make sure to rotate to a new index.
      Defaults to `yyyy.mm`. Uses [dateformat][].
    - `timestampFieldname` the field name of the timestamp. Defaults to
      `@timestamp`.
    - `getTime` a function returning the current time. The return value is
      passed to `new Date(time)`. Default to `Date.now()`.
    - `additionalFields` object with additional fields to be included for each
      metric
    - `pingTimeout` number of millis before the initial ping HEAD request times
      out. The default is the elasticsearch client default (currently `100`).
- `sendBulk()` sends a bulk update using the elasticsearch client
- `start([interval[, unit]])` performs a ping request and once successful sends
  bulk updates every `interval * unit` where `interval` defaults to `60` and
  `unit` defaults to `1000`.
- `stop()` stops performing ping and bulk update requests and calls `end()` on
  all registered collections

## Events

The `reporter` instance returned by `forClient` is an EventEmitter that emits
these events:

- `start` when `start()` was called and the initial ping request was successful
- `stop` when `stop()` was called
- `update` after a bulk update was sent to elasticsearch
- `error` if the initial ping requests fail and if bulk updates fail

[java]: https://github.com/elasticsearch/elasticsearch-metrics-reporter-java
[measured]: https://github.com/felixge/node-measured
[elasticsearch]: https://www.npmjs.com/package/elasticsearch
[dateformat]: https://www.npmjs.com/package/dateformat


CHANGES.md
# Changes

## 1.6.1

- Use ISO time to generate index name to match with `@timestamp`

## 1.6.0

- Allow to configure a `getTime` function that returns the time to use

## 1.5.0

- Move the index to the top level instead of repeating it on every entry

## 1.4.0

- Allow to override the default ping request timeout
- Emit "error" events if ping requests fail

## 1.3.1

- Send `0` values if histogram was never updated

## 1.3.0

- Emit "error" event if bulk update to elasticsearch failed

## 1.2.1

- Fix `additionalFields` for `gauge` metrics
- Disable max listeners warning

## 1.2.0

- Emit "start", "stop" and "update" events

## 1.1.0

- Add support for `additionalFields`

## 1.0.0

- Initial release


test/start-test.js
/*global describe, it, beforeEach, afterEach*/
/*
 * measured-elasticsearch
 *
 * Copyright (c) 2015 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
'use strict';

var assert   = require('assert');
var sinon    = require('sinon');
var measured = require('measured');
var defaults = require('./fixture/defaults');
var api      = require('..');


describe('start', function () {
  var client;
  var reporter;
  var clock;

  beforeEach(function () {
    clock    = sinon.useFakeTimers();
    client   = { ping: sinon.stub().yields(null), bulk: sinon.stub() };
    reporter = api.forClient(client);
  });

  afterEach(function () {
    clock.restore();
  });

  it('sends an initial ping request', function () {
    client.ping.reset();

    reporter.start();

    sinon.assert.calledOnce(client.ping);
  });

  it('retries the ping request every 5 seconds if it failed', function () {
    reporter.on('error', function () { return; });
    client.ping.yields(new Error());

    reporter.start();

    clock.tick(5000);
    clock.tick(5000);

    sinon.assert.calledThrice(client.ping);
  });

  it('does not retry the ping requets if it succeeded', function () {
    reporter.start();

    clock.tick(5000);
    clock.tick(5000);

    sinon.assert.calledOnce(client.ping);
  });

  it('performs a bulk update immediately following the ping', function () {
    var collection = measured.createCollection();
    reporter.addCollection(collection);
    collection.counter('mycount').inc();

    reporter.start();

    sinon.assert.calledOnce(client.bulk);
  });

  it('performs a bulk update every 60 seconds', function () {
    var collection = measured.createCollection();
    reporter.addCollection(collection);
    collection.counter('mycount').inc();

    reporter.start();
    clock.tick(60000);
    clock.tick(60000);

    sinon.assert.calledThrice(client.bulk);
  });

  it('performs a bulk update every n configured seconds', function () {
    reporter.start(10);

    clock.tick(10000);

    sinon.assert.calledTwice(client.bulk);
  });

  it('performs a bulk update every n configured time units', function () {
    reporter.start(500, measured.units.MILLISECONDS);

    clock.tick(500);

    sinon.assert.calledTwice(client.bulk);
  });

  it('does not send updates from removed collections', function () {
    var collection = measured.createCollection();
    reporter.addCollection(collection);
    collection.meter('m').mark();
    reporter.removeCollection(collection);

    reporter.start();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index : sinon.match.string,
      body  : []
    });
  });

  it('stops sending updates', function () {
    var collection = measured.createCollection();
    reporter.addCollection(collection);
    collection.counter('mycount').inc();

    reporter.start();
    client.bulk.reset();
    reporter.stop();
    clock.tick(60000);

    sinon.assert.notCalled(client.bulk);
  });

  it('stops sending pings', function () {
    reporter.on('error', function () { return; });
    client.ping.yields(new Error());

    reporter.start();
    reporter.stop();
    client.ping.reset();
    clock.tick(5000);

    sinon.assert.notCalled(client.ping);
  });

  it('calls end on collections', sinon.test(function () {
    var c1 = measured.createCollection();
    var c2 = measured.createCollection();
    reporter.addCollection(c1);
    reporter.addCollection(c2);
    this.stub(c1, 'end');
    this.stub(c2, 'end');

    reporter.stop();

    sinon.assert.calledOnce(c1.end);
    sinon.assert.calledOnce(c2.end);
  }));

  it('prefixes metric names with collection names', function () {
    var collection = measured.createCollection('foo');
    reporter.addCollection(collection);
    collection.meter('bar').mark();

    reporter.start();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index : sinon.match.string,
      body  : [sinon.match.object, sinon.match.has('name', 'foo.bar')]
    });
  });

  it('emits "start" event before initial update', function () {
    var spy = sinon.spy();
    reporter.on('start', spy);

    reporter.start();

    sinon.assert.calledOnce(spy);
    sinon.assert.callOrder(spy, client.bulk);
  });

  it('does not emits "start" event if ping errs', function () {
    reporter.on('error', function () { return; });
    client.ping.yields(new Error());
    var spy = sinon.spy();
    reporter.on('start', spy);

    reporter.start();

    sinon.assert.notCalled(spy);
  });

  it('emits "update" event after sending initial bulk', function () {
    var spy = sinon.spy();
    reporter.on('update', spy);

    reporter.start();

    sinon.assert.calledOnce(spy);
    sinon.assert.callOrder(client.bulk, spy);
  });

  it('emits "stop" event on stop', function () {
    var spy = sinon.spy();
    reporter.on('stop', spy);

    reporter.stop();

    sinon.assert.calledOnce(spy);
  });

  it('emits "error" event on bulk error', function () {
    var spy = sinon.spy();
    reporter.on('error', spy);
    var err = new Error();
    client.bulk.yields(err);

    reporter.start();

    sinon.assert.calledOnce(spy);
    sinon.assert.calledWith(spy, err);
  });

  it('emits "error" event if ping fails', function () {
    var spy = sinon.spy();
    reporter.on('error', spy);
    var err = new Error();
    client.ping.yields(err);

    reporter.start();

    sinon.assert.calledOnce(spy);
    sinon.assert.calledWith(spy, err);
  });

});


test/config-test.js
/*global describe, it, beforeEach, afterEach*/
/*
 * measured-elasticsearch
 *
 * Copyright (c) 2015 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
'use strict';

var assert   = require('assert');
var sinon    = require('sinon');
var measured = require('measured');
var defaults = require('./fixture/defaults');
var api      = require('..');


describe('config', function () {
  var client;
  var reporter;
  var collection;
  var clock;

  beforeEach(function () {
    clock      = sinon.useFakeTimers();
    client     = { ping: sinon.stub(), bulk: sinon.stub() };
    collection = measured.createCollection();
    collection.meter('foo').mark();
  });

  afterEach(function () {
    clock.restore();
  });

  it('changes default metric name', function () {
    reporter = api.forClient(client, {
      index : 'yo'
    });
    reporter.addCollection(collection);

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index   : 'yo-1970.01',
      body    : [{
        index : sinon.match.object
      }, sinon.match.object]
    });
  });

  it('changes default index date format', function () {
    reporter = api.forClient(client, {
      indexDateFormat : 'yyyy.mm.dd'
    });
    reporter.addCollection(collection);

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index   : 'metrics-1970.01.01',
      body    : [{
        index : sinon.match.object
      }, sinon.match.object]
    });
  });

  it('changes default timestamp name', function () {
    reporter = api.forClient(client, {
      timestampFieldname : 'tick'
    });
    reporter.addCollection(collection);

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index : sinon.match.string,
      body  : [sinon.match.object, sinon.match.has('tick', defaults.timestamp)]
    });
  });

  it('adds configured additional fields to meter', function () {
    reporter = api.forClient(client, {
      additionalFields : {
        server   : 'cheesy-server-name',
        instance : 1
      }
    });
    reporter.addCollection(collection);

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index : sinon.match.string,
      body  : [sinon.match.object, sinon.match({
        server   : 'cheesy-server-name',
        instance : 1
      })]
    });
  });

  it('adds configured additional fields to gauge', function () {
    reporter = api.forClient(client, {
      additionalFields : {
        server   : 'cheesy-server-name',
        instance : 1
      }
    });
    reporter.addCollection(collection);
    collection.gauge('bar', function () {
      return { my : 'custom' };
    });

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index : sinon.match.string,
      body  : [sinon.match.object, sinon.match.object, sinon.match.object,
        sinon.match({
          server   : 'cheesy-server-name',
          instance : 1,
          my       : 'custom'
        })]
    });
  });

  it('uses configured ping request timeout', function () {
    reporter = api.forClient(client, {
      pingTimeout : 12345
    });

    reporter.start();

    sinon.assert.calledOnce(client.ping);
    sinon.assert.calledWith(client.ping, { requestTimeout : 12345 });
  });

  it('does not use a ping request timeout if not configured', function () {
    reporter = api.forClient(client, {});

    reporter.start();

    sinon.assert.calledOnce(client.ping);
    sinon.assert.neverCalledWith(client.ping,
        sinon.match.has('requestTimeout'));
  });

  it('uses configured getTime() function to retrieve the time', function () {
    var getTime = sinon.stub().returns('2015-08-20T12:00:00Z');
    reporter = api.forClient(client, {
      indexDateFormat : 'yyyy.mm.dd',
      getTime         : getTime
    });

    reporter.sendBulk();

    sinon.assert.calledOnce(getTime);
    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index : "metrics-2015.08.20",
      body  : []
    });
  });

  it('uses ISO time to generate index to match with @timestamp', function () {
    var getTime = sinon.stub().returns('2015-08-20T01:00:00+02:00');
    reporter = api.forClient(client, {
      indexDateFormat : 'yyyy.mm.dd',
      getTime         : getTime
    });

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index : "metrics-2015.08.19",
      body  : []
    });
  });

});


test/metrics-test.js
/*global describe, it, beforeEach, afterEach*/
/*
 * measured-elasticsearch
 *
 * Copyright (c) 2015 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
'use strict';

var assert   = require('assert');
var sinon    = require('sinon');
var measured = require('measured');
var defaults = require('./fixture/defaults');
var api      = require('..');


describe('metrics', function () {
  var client;
  var reporter;
  var collection;
  var clock;

  beforeEach(function () {
    clock      = sinon.useFakeTimers();
    client     = { ping: sinon.stub(), bulk: sinon.stub() };
    reporter   = api.forClient(client);
    collection = measured.createCollection();
    reporter.addCollection(collection);
  });

  afterEach(function () {
    clock.restore();
  });

  it('timer stopwatch sends the correct json format', function () {
    var timer = collection.timer('mytime');
    var stopwatch = timer.start();
    clock.tick(1000);
    stopwatch.end();

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index            : defaults.index,
      body             : [defaults.headerTimer, {
        name           : 'mytime',
        '@timestamp'   :  '1970-01-01T00:00:01.000Z',
        count          : 1,
        max            : sinon.match.number,
        mean           : sinon.match.number,
        min            : sinon.match.number,
        p50            : sinon.match.number,
        p75            : sinon.match.number,
        p95            : sinon.match.number,
        p99            : sinon.match.number,
        p999           : sinon.match.number,
        stddev         : 0,
        m1_rate        : 0,
        m5_rate        : 0,
        m15_rate       : 0,
        mean_rate      : sinon.match.number,
        duration_units : 'milliseconds',
        rate_units     : 'calls/second'
      }]
    });
  });

  it('timer meter sends the correct json format', function () {
    var timer = collection.timer('mytime');
    timer.update(5);
    clock.tick(1000);
    timer.update(3);

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index            : defaults.index,
      body             : [defaults.headerTimer, {
        name           : 'mytime',
        '@timestamp'   : '1970-01-01T00:00:01.000Z',
        count          : 2,
        max            : 5,
        mean           : 4,
        min            : 3,
        p50            : 4,
        p75            : 5,
        p95            : 5,
        p99            : 5,
        p999           : 5,
        stddev         : sinon.match.number,
        m1_rate        : 0,
        m5_rate        : 0,
        m15_rate       : 0,
        mean_rate      : sinon.match.number,
        duration_units : 'milliseconds',
        rate_units     : 'calls/second'
      }]
    });
  });

  it('meter sends the correct json format', function () {
    var meter = collection.meter('mymeter');
    meter.mark();
    clock.tick(1000);
    meter.mark();
    clock.tick(1000);
    meter.mark();

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index          : defaults.index,
      body           : [defaults.headerMeter, {
        name         : 'mymeter',
        '@timestamp' : '1970-01-01T00:00:02.000Z',
        count        : 3,
        m1_rate      : 0,
        m5_rate      : 0,
        m15_rate     : 0,
        mean_rate    : sinon.match.number,
        units        : 'events/second'
      }]
    });
  });

  it('meter sends 0 values if mark was never called', function () {
    collection.meter('mymeter');

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index          : defaults.index,
      body           : [defaults.headerMeter, {
        name         : 'mymeter',
        '@timestamp' : '1970-01-01T00:00:00.000Z',
        count        : 0,
        m1_rate      : 0,
        m5_rate      : 0,
        m15_rate     : 0,
        mean_rate    : sinon.match.number,
        units        : 'events/second'
      }]
    });
  });

  it('histogram sends the correct json format', function () {
    var histogram = collection.histogram('myhistogram');
    histogram.update(1);
    clock.tick(1000);
    histogram.update(3);
    clock.tick(1000);
    histogram.update(5);

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index          : defaults.index,
      body           : [defaults.headerHistogram, {
        name         : 'myhistogram',
        '@timestamp' : '1970-01-01T00:00:02.000Z',
        count        : 3,
        max          : 5,
        mean         : 3,
        min          : 1,
        p50          : 3,
        p75          : 5,
        p95          : 5,
        p99          : 5,
        p999         : 5,
        stddev       : 2
      }]
    });
  });

  it('histogram sends 0 values if update was never called', function () {
    collection.histogram('myhistogram');

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index          : defaults.index,
      body           : [defaults.headerHistogram, {
        name         : 'myhistogram',
        '@timestamp' : '1970-01-01T00:00:00.000Z',
        count        : 0,
        max          : 0,
        mean         : 0,
        min          : 0,
        p50          : 0,
        p75          : 0,
        p95          : 0,
        p99          : 0,
        p999         : 0,
        stddev       : 0
      }]
    });
  });

  it('counter sends the correct json format', function () {
    collection.counter('mycount').inc();

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index          : defaults.index,
      body           : [defaults.headerCounter, {
        name         : 'mycount',
        '@timestamp' : defaults.timestamp,
        count        : 1
      }]
    });
  });

  it('counter sends 0 count if inc was never called', function () {
    collection.counter('mycount');

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index          : defaults.index,
      body           : [defaults.headerCounter, {
        name         : 'mycount',
        '@timestamp' : defaults.timestamp,
        count        : 0
      }]
    });
  });

  it('gauge sends the correct json format', function () {
    collection.gauge('mygauge', function () {
      return { random : 42 };
    });

    reporter.sendBulk();

    sinon.assert.calledOnce(client.bulk);
    sinon.assert.calledWith(client.bulk, {
      index          : defaults.index,
      body           : [defaults.headerGauge, {
        name         : 'mygauge',
        '@timestamp' : defaults.timestamp,
        random       : 42
      }]
    });
  });

});


lib/reporter.js
/*
 * measured-elasticsearch
 *
 * Copyright (c) 2015 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
'use strict';

var events     = require('events');
var dateFormat = require("dateformat");

function two(num) {
  return num < 10 ? '0' + num : num;
}

function translateMeter(meter, data) {
  data.count     = meter.count;
  data.m1_rate   = meter['1MinuteRate'];
  data.m5_rate   = meter['5MinuteRate'];
  data.m15_rate  = meter['15MinuteRate'];
  data.mean_rate = meter.mean;
}

function translateHistogram(histogram, data) {
  data.count  = histogram.count;
  data.max    = histogram.max || 0;
  data.mean   = histogram.mean || 0;
  data.min    = histogram.min || 0;
  data.p50    = histogram.median || 0;
  data.p75    = histogram.p75 || 0;
  data.p95    = histogram.p95 || 0;
  data.p99    = histogram.p99 || 0;
  data.p999   = histogram.p999 || 0;
  data.stddev = histogram.stddev || 0;
}

function sendBulk(client, collections, config, callback) {
  var now       = new Date(config.getTime());
  var indexDate = dateFormat(now, config.indexDateFormat, true);
  var index     = config.index + '-' + indexDate;
  var timestamp = now.toISOString();
  var body      = [];

  collections.forEach(function (collection) {
    var prefix = '';
    var json   = collection.toJSON();
    if (collection.name) {
      prefix = collection.name + '.';
      json   = json[collection.name];
    }
    Object.keys(json).forEach(function (name) {
      var raw   = json[name];
      var data  = JSON.parse(config.additionalFields);
      var type;
      if (typeof raw === 'number') {
        type = 'counter';
        data.count = raw;
      } else if (raw.histogram && raw.meter) {
        type = 'timer';
        translateHistogram(raw.histogram, data);
        translateMeter(raw.meter, data);
        data.duration_units = 'milliseconds';
        data.rate_units     = 'calls/second';
      } else if (raw.hasOwnProperty('1MinuteRate')) {
        type = 'meter';
        translateMeter(raw, data);
        data.units = 'events/second';
      } else if (raw.hasOwnProperty('p999')) {
        type = 'histogram';
        translateHistogram(raw, data);
      } else {
        type = 'gauge';
        Object.keys(raw).forEach(function (key) {
          data[key] = raw[key];
        });
      }
      body.push({
        index   : {
          _type : type
        }
      });
      data.name = prefix + name;
      data[config.timestampFieldname] = timestamp;
      body.push(data);
    });
  });

  client.bulk({ index : index, body : body }, callback);
}

function endCollection(collection) {
  collection.end();
}

function getTime() {
  return Date.now();
}


exports.forClient = function (client, config) {
  if (!config) {
    config = {};
  }
  if (!config.index) {
    config.index = 'metrics';
  }
  if (!config.indexDateFormat) {
    config.indexDateFormat = 'yyyy.mm';
  }
  if (!config.timestampFieldname) {
    config.timestampFieldname = '@timestamp';
  }
  if (!config.additionalFields) {
    config.additionalFields = {};
  }
  if (!config.getTime) {
    config.getTime = getTime;
  }
  config.additionalFields = JSON.stringify(config.additionalFields);

  var collections    = [];
  var updateInterval = null;
  var pingTimeout    = null;
  var reporter       = new events.EventEmitter();
  reporter.setMaxListeners(0);

  function update() {
    sendBulk(client, collections, config, function (err) {
      if (err) {
        reporter.emit('error', err);
      }
    });
    reporter.emit('update');
  }

  reporter.sendBulk = update;

  var pingParams = {};
  if (config.pingTimeout) {
    pingParams.requestTimeout = config.pingTimeout;
  }
  reporter.start = function start(interval, unit) {
    client.ping(pingParams, function (err) {
      if (err) {
        reporter.emit('error', err);
        pingTimeout = setTimeout(start, 5000);
        return;
      }
      pingTimeout = null;
      reporter.emit('start');
      update();
      var updateMillis = (interval || 60) * (unit || 1000);
      updateInterval = setInterval(update, updateMillis);
    });
  };

  reporter.stop = function () {
    if (pingTimeout !== null) {
      clearTimeout(pingTimeout);
      pingTimeout = null;
    }
    if (updateInterval !== null) {
      clearInterval(updateInterval);
      updateInterval = null;
    }
    collections.forEach(endCollection);
    reporter.emit('stop');
  };

  reporter.addCollection = function (collection) {
    collections.push(collection);
  };

  reporter.removeCollection = function (collection) {
    var p = collections.indexOf(collection);
    if (p !== -1) {
      collections.splice(p, 1);
    }
  };

  return reporter;
};
